### PR TITLE
GEOMESA-3201 Support configuration of parallel partition scans in dat…

### PR DESCRIPTION
**Pull request by elahrvivaz**
…a store parameters

### CODE REVIEW

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/util/BatchMultiScanner.scala [hunk 1]
```diff
@@ -98,10 +101,10 @@ class BatchMultiScanner(
           val entry = inQ.poll(5, TimeUnit.MILLISECONDS)
           if (entry != null) {
             val entries = collection.mutable.ListBuffer(entry)
-            inQ.drainTo(entries)
+            inQ.drainTo(entries.asJava)
             val task = executor.submit(new Runnable {
               override def run(): Unit = {
-                val iterator = join.copy(ranges = entries.map(joinFunction)).scan(connector, auths, timeout)
+                val iterator = join.copy(ranges = entries.map(joinFunction)).scan(connector, auths, partitionParallelScans, timeout)
```
Comment (jnh5y) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/util/BatchMultiScanner.scala:107]: What is this coming to again?

Comment (elahrvivaz) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/util/BatchMultiScanner.scala:107]: this is just passing the parallel flag back into the query plan scan here: https://github.com/locationtech/geomesa/pull/2860/files#diff-db1291fc355346779b080ab328128d55e40f509c9372aa183b54a889511a2f71R127

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedFeatureSourceView.scala [hunk 2]
```diff
@@ -46,11 +48,18 @@ class MergedFeatureSourceView(
...
   override def getCount(query: Query): Int = {
-    // if one of our sources can't get a count (i.e. is negative), give up and return -1
-    val total = sources.foldLeft(0) { case (sum, (source, filter)) =>
-      lazy val count = source.getCount(mergeFilter(query, filter))
-      if (sum < 0 || count < 0) { -1 } else { sum + count }
-    }
+    val total =
+      if (parallel) {
+        val counts = sources.par.map { case (source, f) => source.getCount(mergeFilter(query, f)) }
+        counts.foldLeft(0)((sum, count) => if (sum < 0 || count < 0) { -1 } else { sum + count })
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedFeatureSourceView.scala:54]: The only unfortunate thing about this parallel computation is that all the counts will be computed even if one of the sources returns a -1, right?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedQueryRunner.scala [hunk 3]
```diff
@@ -254,6 +266,16 @@ class MergedQueryRunner(ds: HasGeoMesaStats, stores: Seq[(Queryable, Option[Filt
       super.getReturnSft(sft, hints)
     }
   }
+
+  private def doParallelScan(
+      readers: Seq[FeatureReader[SimpleFeatureType, SimpleFeature]],
+      single: FeatureReader[SimpleFeatureType, SimpleFeature] => CloseableIterator[SimpleFeature]): CloseableIterator[SimpleFeature] = {
+    if (parallel) {
+      SelfClosingIterator(readers.par.map(single).iterator).flatMap(i => i)
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedQueryRunner.scala:274]: Do the readers need to be closed?  Does this do that?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedFeatureSourceView.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedFeatureSourceView.scala:54]: yeah... we could map them to futures and try to cancel them when we see a -1, but i think that would likely not work great given all the effort we had to put in to get query timeouts working.

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedFeatureSourceView.scala [hunk 4]
```diff
@@ -59,23 +68,28 @@ class MergedFeatureSourceView(
...
-      case (source, None)    => bounds.expandToInclude(source.getBounds)
-      case (source, Some(f)) => bounds.expandToInclude(source.getBounds(new Query(sft.getTypeName, f)))
+    def getSingle(sourceAndFilter: (SimpleFeatureSource, Option[Filter])): Option[ReferencedEnvelope] = {
+      sourceAndFilter match {
+        case (source, None)    => Option(source.getBounds)
+        case (source, Some(f)) => Option(source.getBounds(new Query(sft.getTypeName, f)))
+      }
     }
+
+    val sourceBounds = if (parallel) { sources.par.flatMap(getSingle).seq } else { sources.flatMap(getSingle) }
```
[the remark this answers is not in the document]
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedFeatureSourceView.scala:78]: yeah, similar to the count method, we'd have to either not parallelize everything, or try to cancel some futures...

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedQueryRunner.scala (hunk 3 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedQueryRunner.scala:274]: i was hoping you'd confirm haha. i _think_ the flatMap(i => i) will take care of it.

Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedQueryRunner.scala:274]: `flatten` seemed a bit more complicated to implement for closeable iterators

Comment (jnh5y) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedQueryRunner.scala:274]: Honestly, this stuff is complicated enough that I'd end up writing a unit test to convince myself that I didn't screw up...

And later, I'd find that I missed something.:)

**Review by jnh5y — APPROVED** [review]
:shipit: !

The only thing to consider is adding a test to cover the discussion here: https://github.com/locationtech/geomesa/pull/2860#discussion_r894662933

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedQueryRunner.scala (hunk 3 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedQueryRunner.scala:280]: added some tests

Comment (jnh5y) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/MergedQueryRunner.scala:280]: Ship it when ready!  (Or tell me to push the button.)

**Pull request merged**